src/host.cpp

```diff
@@ -41,4 +35,34 @@ void remove_host(CopyOnWriteHostVec& hosts, const SharedRefPtr<Host>& host) {
   }
 }
 
+void Host::LatencyTracker::update(uint64_t latency_ns) {
+  uint64_t now = uv_hrtime();
+
+  ScopedSpinlock l(SpinlockPool<LatencyTracker>::get_spinlock(this));
+
+  TimestampedAverage previous = current_;
+
+  if (previous.num_measured < threshold_to_account_) {
+    current_.average = -1;
+    current_.num_measured = previous.num_measured + 1;
+    current_.timestamp = now;
+  } else if (previous.average < 0) {
+    current_.average = latency_ns;
+    current_.num_measured = previous.num_measured + 1;
+    current_.timestamp = now;
+  } else {
+    int64_t delay = now - previous.timestamp;
+    if (delay <= 0) {
+      return;
+    }
+
+    double scaled_delay = static_cast<double>(delay) / scale_ns_;
+    double weight = log(scaled_delay + 1) / scaled_delay;
+
+    current_.average = static_cast<int64_t>((1.0 - weight) * latency_ns + weight * previous.average);
```

Comment:
My main concern was based on a misreading of the weight calculation. I'm still not sure about the derivation, but it seems like it should behave fine.